pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TransactionBufferHandlerImpl.java

```diff
@@ -113,24 +110,38 @@ public TransactionBufferHandlerImpl(ConnectionPool connectionPool, NamespaceServ
         long requestId = requestIdGenerator.getAndIncrement();
         ByteBuf cmd = Commands.newEndTxnOnSubscription(requestId, txnIdLeastBits, txnIdMostBits,
                 topic, subscription, action, lowWaterMark);
+        return endTxn(requestId, topic, cmd, cb);
+    }
+
+    private CompletableFuture<TxnID> endTxn(long requestId, String topic, ByteBuf cmd, CompletableFuture<TxnID> cb) {
         OpRequestSend op = OpRequestSend.create(requestId, topic, cmd, cb);
         pendingRequests.put(requestId, op);
         cmd.retain();
-        cnx(topic).whenComplete((clientCnx, throwable) -> {
-            if (throwable == null) {
-                try {
-                    clientCnx.ctx().writeAndFlush(cmd, clientCnx.ctx().voidPromise());
-                } catch (Exception e) {
-                    cb.completeExceptionally(e);
+        try {
+            cache.get(topic).whenComplete((clientCnx, throwable) -> {
+                if (throwable == null) {
+                    try {
+                        clientCnx.registerTransactionBufferHandler(TransactionBufferHandlerImpl.this);
+                        clientCnx.ctx().writeAndFlush(cmd, clientCnx.ctx().voidPromise());
+                    } catch (Exception e) {
+                        cache.refresh(topic);
+                        cb.completeExceptionally(e);
+                        pendingRequests.remove(requestId);
+                        op.recycle();
+                    }
+                } else {
+                    cache.refresh(topic);
```

Comment:
is it better to use "refresh" or "invalidate" here ?

refresh tries to load a new value, asynchronously, if we received an error, it is probable that we will see an error during refresh as well
https://guava.dev/releases/19.0/api/docs/com/google/common/cache/LoadingCache.html#refresh(K)

Reply:
"invalidate" is better, i will change. good check thanks.